cmd/kube-proxy/app/server.go

```diff
@@ -205,7 +205,7 @@ func NewProxyServerDefault(config *options.ProxyServerConfig) (*ProxyServer, err
 			return nil, fmt.Errorf("Unable to read IPTablesMasqueradeBit from config")
 		}
 
-		proxierIptables, err := iptables.NewProxier(iptInterface, execer, config.IPTablesSyncPeriod.Duration, config.MasqueradeAll, int(*config.IPTablesMasqueradeBit), config.ClusterCIDR)
+		proxierIptables, err := iptables.NewProxier(iptInterface, execer, config.IPTablesSyncPeriod.Duration, config.MasqueradeAll, int(*config.IPTablesMasqueradeBit), config.ClusterCIDR, hostname)
```

Comment:
I believe we want the hostname here to match the kubelet's nodename. The kubelet's nodename == hostname if there's no hostname override, so you got that part right, however I think the kubelet will prefer the cloudprovider specified name over os.Hostname, which means that's the name in the pod.Spec, which makes that the name in the endpoints object. 

Do we need to refactor this function: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L654 ? I guess kube-proxy doesn't understand cloudprovider, so maybe we should stay away from trying to compute hostname and just wait for kubelet to create the node and get the name from there? 

I haven't tracked the rathole that is kubelet.NodeName/HostName, it might be that my concerns are invalid.